[jit] support dict augment assign in script

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23842 [jit] support grad and data attribute for tensor in script
* #23841 [jit] add initial support for sparse tensors
* **#23639 [jit] support dict augment assign in script**

Summary:
Added the dict augmented assignment to TorchScript, which allows
something like:

```
def test(a):
    # type: (Dict[str, int]) -> Dict[str, int]
    a['a'] += 1
    return a
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23639

Differential Revision: [D16683559](https://our.internmc.facebook.com/intern/diff/D16683559)

wanchaol committed:
[jit] support dict augment assign in script

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] support dict augment assign in script"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] support dict augment assign in script"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

wanchaol committed:
Update on "[jit] support dict augment assign in script"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D16683559](https://our.internmc.facebook.com/intern/diff/D16683559)

suo commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1657,6 +1657,38 @@ struct to_ir {
   }
 
+  void emitAugAssignToContainer(
+      const std::string& containerType,
```
Passing the container type as a list is very suspicious. Can we pass the actual `Type` and use `python_str()` methods and such for reporting errors?

suo commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1706,36 +1738,22 @@ struct to_ir {
-          getAugOp(stmt, elementType), {getItem, valueArg}, {}, stmt.range());
-      graph->insert(
-          aten::_set_item, {listArg, idxArg, augmentedItem}, {}, stmt.range());
+      emitAugAssignToContainer("list", elementType, stmt, lhs, sliceable);
```
Similarly, you don't need this branching here—it can just be `emitAugAssignGeneric` and take a type. Internally that function can do the check that the type is supported

suo reviewed (changes requested):
meant to request changes

wanchaol committed:
Update on "[jit] support dict augment assign in script"


Summary:
Added the dict augmented assignment to TorchScript, which allows
something like:

```
def test(a):
    # type: (Dict[str, int]) -> Dict[str, int]
    a['a'] += 1
    return a
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23639

Differential Revision: [D16683559](https://our.internmc.facebook.com/intern/diff/D16683559)

suo commented on torch/csrc/jit/script/compiler.cpp:
```diff
@@ -1683,6 +1683,45 @@ struct to_ir {
+      elemType = dictType->getKeyType();
+    }
+
+    if (subscriptExprs.size() != 1 || elemType == nullptr) {
```
These are actually two different errors. The lhs is "Slices are not supported". The rhs is "this type doesn't support augmented assignment at all". Can we differentiate between these two?

suo reviewed (approved):
minor nit but otherwise lgtm

wanchaol committed:
rebase and address comments on "[jit] support dict augment assign in script"


Summary:
Added the dict augmented assignment to TorchScript, which allows
something like:

```
def test(a):
    # type: (Dict[str, int]) -> Dict[str, int]
    a['a'] += 1
    return a
```

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Pull Request resolved: https://github.com/pytorch/pytorch/pull/23639

Differential Revision: [D16683559](https://our.internmc.facebook.com/intern/diff/D16683559)
